raft/rawnode_test.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writinrawNode, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package raft
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// TestRawNodeStep ensures that RawNode.Step ignore local message.
+func TestRawNodeStep(t *testing.T) {
+	for i, msgn := range raftpb.MessageType_name {
+		s := NewMemoryStorage()
+		rawNode, _ := NewRawNode(newTestConfig(1, nil, 10, 1, s), []Peer{{ID: 1}})
+		msgt := raftpb.MessageType(i)
+		err := rawNode.Step(raftpb.Message{Type: msgt})
+		// LocalMsg should be ignored.
+		if msgt == raftpb.MsgBeat || msgt == raftpb.MsgHup || msgt == raftpb.MsgUnreachable || msgt == raftpb.MsgSnapStatus {
+			if err != nil {
+				t.Errorf("%d: step should ignore %s", msgt, msgn)
+			}
+		}
+	}
+}
+
+// TestRawNodePropose ensures that RawNode.Propose sends the given proposal to the underlying raft.
+func TestRawNodePropose(t *testing.T) {
+	s := NewMemoryStorage()
+	rawNode, _ := NewRawNode(newTestConfig(1, nil, 10, 1, s), []Peer{{ID: 1}})
+	rawNode.Campaign()
+	proposed := false
+	for {
+		rd := rawNode.Ready()
+		s.Append(rd.Entries)
+		// Once we are the leader, propose a command.
+		if !proposed && rd.SoftState.Lead == rawNode.raft.id {
+			rawNode.Propose([]byte("somedata"))
+			proposed = true
+		}
+		rawNode.Advance(rd)
+
+		// Exit when we have three entries: one ConfChange, one no-op for the election,
+		// and our proposed command.
+		lastIndex, err := s.LastIndex()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if lastIndex >= 3 {
+			break
+		}
+	}
+
+	lastIndex, err := s.LastIndex()
+	if err != nil {
+		t.Fatal(err)
+	}
+	entries, err := s.Entries(lastIndex, lastIndex+1, noLimit)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(entries) != 1 {
+		t.Fatalf("len(entries) = %d, want %d", len(entries), 1)
+	}
+	if !bytes.Equal(entries[0].Data, []byte("somedata")) {
+		t.Errorf("entries[0].Data = %v, want %v", entries[0].Data, []byte("somedata"))
+	}
+}
+
+// TestRawNodeProposeConfig ensures that RawNode.ProposeConfChange
+// sends the given configuration proposal to the underlying raft.
+func TestRawNodeProposeConfig(t *testing.T) {
```

Comment:
It seems to be easy enough to combine this test and the propose one with a for loop. But it is also fine to keep both.